src/client/client.cpp

```diff
@@ -129,10 +130,33 @@ Client::Client(
 	m_env.setLocalPlayer(new LocalPlayer(this, playername));
 
 	// Make the mod storage database and begin the save for later
-	m_mod_storage_database =
-		new ModMetadataDatabaseSQLite3(porting::path_user + DIR_DELIM + "client");
+	std::string mod_storage_dir = porting::path_user + DIR_DELIM + "client";
+	m_mod_storage_database = new ModMetadataDatabaseSQLite3(mod_storage_dir);
 	m_mod_storage_database->beginSave();
 
+	// Migrate old files-based mod storage
+	std::string old_mod_storage = mod_storage_dir + DIR_DELIM + "mod_storage";
+	if (fs::IsDir(old_mod_storage)) {
+		infostream << "Migrating client mod storage to SQLite3 database" << std::endl;
+		{
+			ModMetadataDatabaseFiles files_db(mod_storage_dir);
+			std::vector<std::string> mod_list;
+			files_db.listMods(&mod_list);
+			for (const std::string &modname : mod_list) {
+				infostream << "Migrating client mod storage for mod " << modname << std::endl;
+				StringMap meta;
+				files_db.getModEntries(modname, &meta);
+				for (const auto &pair : meta) {
+					m_mod_storage_database->setModEntry(modname, pair.first, pair.second);
+				}
+			}
+		}
+		if (!fs::Rename(old_mod_storage, old_mod_storage + ".bak"))
+			// Execution cannot move forward if the migration does not complete.
+			throw BaseException("Could not finish migrating client mod storage");
```

Comment:
I'm not a big fan of this pattern, it would be better to either move the comment up a line or introduce braces here for clarity.